hazelcast/src/main/java/com/hazelcast/util/collection/InflatableSet.java

```diff
@@ -0,0 +1,251 @@
+
+        //Only array-backed representation exists
+        COMPACT,
+
+        //both array-backed & hashset-backed representation exist.
+        //this is needed as we are creating HashSet on contains()
+        //but we don't want invalidate existing iterators.
+        HYBRID,
+
+        //only hashset based representation exists
+        INFLATED
+    }
+
+    private final List<T> compactList;
+    private Set<T> inflatedSet;
+    private State state;
+
+    public InflatableSet(int initialCapacity) {
+        this.state = State.INITIAL_LOAD;
+        if (initialCapacity < 0) {
+            throw new IllegalArgumentException("Initial Capacity must be a positive number. Current capacity: "
+                    + initialCapacity);
+        }
+        compactList = new ArrayList<T>(initialCapacity);
+    }
+
+    protected InflatableSet(InflatableSet other) {
+        compactList = new ArrayList<T>(other.compactList.size());
+        compactList.addAll(other.compactList);
+        if (other.inflatedSet != null) {
+            inflatedSet = new HashSet<T>(other.inflatedSet);
+        }
+        state = other.state;
+    }
+
+
+    @Override
+    public int size() {
+        if (state == State.INFLATED) {
+            return inflatedSet.size();
+        } else {
+            return compactList.size();
+        }
+    }
+
+    @Override
+    public boolean isEmpty() {
+        if (state == State.INFLATED) {
+            return inflatedSet.isEmpty();
+        } else {
+            return compactList.isEmpty();
+        }
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        if (state == State.INITIAL_LOAD) {
+            return compactList.contains(o);
+        }
+        if (state == State.COMPACT) {
+            toHybridState();
+        }
+        return inflatedSet.contains(o);
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        if (state == State.INFLATED) {
+            return inflatedSet.iterator();
+        }
+        return new HybridIterator();
+    }
+
+    @Override
+    public boolean add(T t) {
+        if (state == State.INITIAL_LOAD) {
+            return compactList.add(t);
+        }
+        toInflatedState();
+        return inflatedSet.add(t);
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        switch (state) {
+            case HYBRID:
+                compactList.remove(o);
+                return inflatedSet.remove(o);
+            case INFLATED:
+                return inflatedSet.remove(o);
+            default:
+                return compactList.remove(o);
+        }
+    }
+
+    @Override
+    public void clear() {
+        switch (state) {
+            case HYBRID:
+                inflatedSet.clear();
+                compactList.clear();
+                break;
+            case INFLATED:
+                inflatedSet.clear();
+                break;
+            default:
+                compactList.clear();
+        }
+    }
+
+    public void close() {
```

Comment:
Closing a collection might be a bit misleading. If I close something I wouldn't expect to add stuff to it but here I can. Perhaps we could use the Builder pattern to construct an InflatableSet and pass it a List as a constructor argument.